src/caret_analyze/architecture/architecture.py

```diff
@@ -365,35 +365,58 @@ def assign_message_context(self, node_name: str, context_type: str,
 
-            node.update_node_path(NodeValuesLoaded._search_node_paths(node, context_reader))
+    def insert_publisher_callback(self, node_name: str,
+                                  publish_topic_name: str, callback_name: str):
+        node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
 
-    def assign_publisher_and_callback(self, node_name: str,
-                                      publish_topic_name: str, callback_name: str):
+        node.insert_publisher_callback(publish_topic_name, callback_name)
+
+        node.update_node_path(NodeValuesLoaded._search_node_paths(node,
+                              AssignContextReader(node)))
+
+    def insert_variable_passing(self, node_name: str,
+                                callback_name_write: str, callback_name_read: str):
         node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
 
-        node.assign_publisher_and_callback(publish_topic_name, callback_name)
+        node.insert_variable_passing(callback_name_write, callback_name_read)
 
         node.update_node_path(NodeValuesLoaded._search_node_paths(node,
                               AssignContextReader(node)))
 
-    def assign_variable_passings(self, node_name: str,
-                                 callback_name_write: str, callback_name_read: str):
+    def remove_publisher_and_callback(self, node_name: str,
+                                      publish_topic_name: str, callback_name: str):
         node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
 
-        node.assign_variable_passings(callback_name_write, callback_name_read)
+        node.remove_publisher_and_callback(publish_topic_name, callback_name)
 
         node.update_node_path(NodeValuesLoaded._search_node_paths(node,
                               AssignContextReader(node)))
 
+    def remove_variable_passing(self, node_name: str,
+                                callback_name_write: str, callback_name_read: str):
+        node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
+
+        node.remove_variable_passing(callback_name_write, callback_name_read)
+
+        callback_read: CallbackStructValue = \
+            Util.find_one(lambda x: x.callback_name == callback_name_read, self.callbacks)
+        callback_write: CallbackStructValue = \
+            Util.find_one(lambda x: x.callback_name == callback_name_write, self.callbacks)
+
+        context_reader = AssignContextReader(node)
+        for publish_topic_name in callback_read.publish_topic_names or []:
+            if callback_write.subscribe_topic_name is not None:
+                context_reader.remove_callback_chain(callback_write.subscribe_topic_name,
+                                                     publish_topic_name)
+
+        node.update_node_path(NodeValuesLoaded._search_node_paths(node,
+                              context_reader))
```

Comment:
If the following implementation does not change the logical output, it can be made more efficient.

```suggestion
        if callback_read.publish_topic_names:
            context_reader = AssignContextReader(node)
            for publish_topic_name in callback_read.publish_topic_names:
                if callback_write.subscribe_topic_name is not None:
                    context_reader.remove_callback_chain(callback_write.subscribe_topic_name,
                                                         publish_topic_name)

            node.update_node_path(NodeValuesLoaded._search_node_paths(node,
                                                                      context_reader))
```

Reply:
Fix it in [67b673a](https://github.com/tier4/CARET_analyze/pull/297/commits/67b673ad8ccaaaaa5f49ae230c3f2f9752062087).